### ISSUE TITLE
Slow `nautilus_trader.common.component.is_matching`

### ISSUE THREAD

**Issue by ryantam626** [nautechsystems/nautilus_trader#2150]
# Bug Report

### Expected Behavior
Expected `nautilus_trader.common.component.is_matching` to be faster.

### Actual Behavior
`nautilus_trader.common.component.is_matching` is prohibitively slow when dealing with instruments with long names (e.g. Polymarket - which results in topic like `data.book.deltas.POLYMARKET.0xf6ea0a571a43a122471e1c7604d6bd8c267082a39758bdd8162dbfae0cd8d137-40739562057794550541187194569338873089092359730180442205123214215534855241164`)

### Steps to Reproduce the Problem

```py
from nautilus_trader.common.component import is_matching_py

topic = 'data.instrument.POLYMARKET.0x4234768b3ad78225ec4cacf078a98044a880d1f0abf51997f22da9670b4a95e4-73174005964727253088253639382577689462665540891931226338645962800718025375512'
pattern = 'data.instrument.POLYMARKET.0x034e00691eb26f6023c35d196e4da700d440e39263957b54ee613b42bac0b9cc-18339512646452261434884166324269813378889037675218398618252803207699512685316'

# ipyton timeit magic
%timeit is_matching_py(topic, pattern)
```

gives
```text
82.7 μs ± 5.7 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

### Further explaination
`82.7 μs` doesn't sound too bad when considered in isolation (also there is further optimisation if we don't cross Cython boundary - let's ignore that for the purpose of this ticket).

But if we want to trade every single instrument in Polymarket, then we could have about 30000+ instruments (this figure was correct when I grabbed all instruments from Polymarket for testing a while back - could be wrong now).

When we want to subscribe to a new topic, we would attempt to match the topic to be subscribed against all patterns in the message bus, of which is filled with long patterns like the above, this would take about `30000 * 82μs` = `2.46s` which is too expensive. 

### Specifications

- OS platform: Windows WSL (`uname -r` output: `5.15.167.4-microsoft-standard-WSL2`)
- Python version: 3.11.10
- `nautilus_trader` version: 1.205.0 (from github and built locally, debug build)

**Comment by ryantam626** [comment]
Proposed fix:

```patch
From 823616a7b1c12f0dc4c654b6809c68a755f7197e Mon Sep 17 00:00:00 2001
From: Ryan Tam <ryantam626@gmail.com>
Date: Sun, 29 Dec 2024 15:50:59 +0000
Subject: [PATCH] Don't use a more expensive O(nm) algorithm if not necessary

Polymarket has excessively long names for instruments[1], the `is_matching`
O(nm) algorithm becomes prohibitively expensive when dealing with
Polymarket instruments - this is further exacerbated when we maintain
all instruments from Polymarket instead of a selected few.

I think there is an alternative optimisation here, which involves
converting the inputs into deterministic finite automata and try to walk
them in sync, however this takes time to implement and validate, and
thus is out of scope right now.

[1] - example of Polymarket related channel/topic:
`data.book.deltas.POLYMARKET.0xf6ea0a571a43a122471e1c7604d6bd8c267082a39758bdd8162dbfae0cd8d137-40739562057794550541187194569338873089092359730180442205123214215534855241164`
---
 nautilus_trader/common/component.pyx | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/nautilus_trader/common/component.pyx b/nautilus_trader/common/component.pyx
index 131db7313..1c59ff01f 100644
--- a/nautilus_trader/common/component.pyx
+++ b/nautilus_trader/common/component.pyx
@@ -2596,7 +2596,17 @@ cdef class MessageBus:
         return subs_array
 
 
+cdef inline bint contains_wildcard(str topic_or_pattern):
+    return '?' in topic_or_pattern or '*' in topic_or_pattern
+
+
 cdef inline bint is_matching(str topic, str pattern):
+    topic_contains_wildcard = contains_wildcard(topic)
+    pattern_contains_wildcard = contains_wildcard(pattern)
+
+    if not topic_contains_wildcard and not pattern_contains_wildcard:
+        return topic == pattern
+
     # Get length of string and wildcard pattern
     cdef int n = len(topic)
     cdef int m = len(pattern)
-- 
2.43.0
```

New performance characteristics:

```pycon
In [1]: from nautilus_trader.common.component import is_matching_py
   ...:
   ...: topic = 'data.instrument.POLYMARKET.0x4234768b3ad78225ec4cacf078a98044a880d1f0abf51997f22da9670b4a95e4-73174005964727253088253639382577689462665540891931226338645962800718025375512'
   ...: pattern = 'data.instrument.POLYMARKET.0x034e00691eb26f6023c35d196e4da700d440e39263957b54ee613b42bac0b9cc-18339512646452261434884166324269813378889037675218398618252803207699512685316'

In [2]: %timeit is_matching_py(topic, pattern)
31.8 ns ± 0.112 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Let me know if you want me to open a PR 🙏

**Comment by faysou** [comment]
I think you can definitely do a PR based on your research above.

**Comment by cjdsellers** [comment]
Hi @ryantam626 

Thank you for reaching out with this contribution!

This all makes sense to me, definitely put up a PR when you can :pray:

**Comment by cjdsellers** [comment]
Now closed from #2151 many thanks @ryantam626
